pkg/controller/oam/applicationconfiguration/render.go

```diff
@@ -0,0 +1,228 @@
+// workloads and traits.
+type ComponentRenderFn func(ctx context.Context, ac *v1alpha2.ApplicationConfiguration) ([]Workload, error)
+
+// Render an ApplicationConfiguration's Components into workloads and traits.
+func (fn ComponentRenderFn) Render(ctx context.Context, ac *v1alpha2.ApplicationConfiguration) ([]Workload, error) {
+	return fn(ctx, ac)
+}
+
+type components struct {
+	client   client.Reader
+	params   ParameterResolver
+	workload ResourceRenderer
+	trait    ResourceRenderer
+}
+
+func (r *components) Render(ctx context.Context, ac *v1alpha2.ApplicationConfiguration) ([]Workload, error) {
+	workloads := make([]Workload, len(ac.Spec.Components))
+	for i, acc := range ac.Spec.Components {
+
+		c := &v1alpha2.Component{}
+		nn := types.NamespacedName{Namespace: ac.GetNamespace(), Name: acc.ComponentName}
+		if err := r.client.Get(ctx, nn, c); err != nil {
+			return nil, errors.Wrapf(err, errFmtGetComponent, acc.ComponentName)
+		}
+
+		p, err := r.params.Resolve(c.Spec.Parameters, ac.Spec.Components[i].ParameterValues)
+		if err != nil {
+			return nil, errors.Wrapf(err, errFmtResolveParams, acc.ComponentName)
+		}
+
+		w, err := r.workload.Render(c.Spec.Workload.Raw, p...)
+		if err != nil {
+			return nil, errors.Wrapf(err, errFmtRenderWorkload, acc.ComponentName)
+		}
+
+		ref := metav1.NewControllerRef(ac, v1alpha2.ApplicationConfigurationGroupVersionKind)
+		w.SetOwnerReferences([]metav1.OwnerReference{*ref})
+		w.SetNamespace(ac.GetNamespace())
+
+		traits := make([]unstructured.Unstructured, len(acc.Traits))
+		for i, ct := range acc.Traits {
+			t, err := r.trait.Render(ct.Trait.Raw)
+			if err != nil {
+				return nil, errors.Wrapf(err, errFmtRenderTrait, acc.ComponentName)
+			}
+
+			ref := metav1.NewControllerRef(ac, v1alpha2.ApplicationConfigurationGroupVersionKind)
+			t.SetOwnerReferences([]metav1.OwnerReference{*ref})
+			t.SetNamespace(ac.GetNamespace())
+
+			traits[i] = *t
+		}
+
+		workloads[i] = Workload{ComponentName: acc.ComponentName, Workload: w, Traits: traits}
+	}
+	return workloads, nil
+}
+
+// A ResourceRenderer renders a Kubernetes-compliant YAML resource into an
+// Unstructured object, optionally setting the supplied parameters.
+type ResourceRenderer interface {
+	Render(data []byte, p ...Parameter) (*unstructured.Unstructured, error)
+}
+
+// A ResourceRenderFn renders a Kubernetes-compliant YAML resource into an
+// Unstructured object, optionally setting the supplied parameters.
+type ResourceRenderFn func(data []byte, p ...Parameter) (*unstructured.Unstructured, error)
+
+// Render the supplied Kubernetes YAML resource.
+func (fn ResourceRenderFn) Render(data []byte, p ...Parameter) (*unstructured.Unstructured, error) {
+	return fn(data, p...)
+}
+
+func renderWorkload(data []byte, p ...Parameter) (*unstructured.Unstructured, error) {
+	// TODO(negz): Is there a better decoder to use here?
+	w := &fieldpath.Paved{}
+	if err := yaml.Unmarshal(data, w); err != nil {
```

Comment:
Values in the field `Raw` of runtime.RawExtension is json bytes I think, I don't know if it works

Reply:
You're right, thanks! I've switched to using `pkg/encoding/json`.